File: python/exercises/code-smells/feature-envy/src/order.py
from dataclasses import dataclass, field
from datetime import datetime
from typing import List, Dict, Any
from customer import Customer
from product import Product


@dataclass
class OrderItem:
    product: Product
    quantity: int


@dataclass
class Order:
    id: str
    customer: Customer
    items: List[OrderItem]
    shipping_address: str
    express: bool = False
    status: str = field(default='pending', init=False)
    order_date: str = field(default_factory=lambda: datetime.now().isoformat(), init=False)

    def get_id(self) -> str:
        return self.id

    def get_customer(self) -> Customer:
        return self.customer

    def get_items(self) -> List[OrderItem]:
        return self.items

    def get_status(self) -> str:
        return self.status

    def get_shipping_address(self) -> str:
        return self.shipping_address

    def get_order_date(self) -> str:
        return self.order_date

    def is_express(self) -> bool:
        return self.express

    def set_status(self, status: str) -> None:
        self.status = status

    def add_item(self, product: Product, quantity: int) -> None:
        self.items.append(OrderItem(product=product, quantity=quantity))

File: typescript/exercises/code-smells/feature-envy/src/Order.ts
import { Customer } from './Customer';
import { Product } from './Product';

export interface OrderItem {
  product: Product;
  quantity: number;
}

export class Order {
  private status: string = 'pending';
  private readonly orderDate: string;

  constructor(
    private readonly id: string,
    private readonly customer: Customer,
    private readonly items: OrderItem[],
    private readonly shippingAddress: string,
    private readonly express: boolean = false
  ) {
    this.orderDate = new Date().toISOString();
  }

  getId(): string {
    return this.id;
  }

  getCustomer(): Customer {
    return this.customer;
  }

  getItems(): OrderItem[] {
    return this.items;
  }

  getStatus(): string {
    return this.status;
  }

  getShippingAddress(): string {
    return this.shippingAddress;
  }

  getOrderDate(): string {
    return this.orderDate;
  }

  isExpress(): boolean {
    return this.express;
  }

  setStatus(status: string): void {
    this.status = status;
  }

  addItem(product: Product, quantity: number): void {
    this.items.push({ product, quantity });
  }
}
